### DISCUSSION TITLE
Load dark mode on static site without first loading light mode, or vice versa

### DISCUSSION THREAD

**Discussion by johnmathews** [tailwindlabs/tailwindcss#3904] [Help]
On a static [site](https://johnmathews.eu) I'm using the dark mode feature introduced in v2.0 with the 'class' method, because I want to let users choose the theme regardless of the OS preference.

This relies on inserting a "dark" class into a DOM element, to make child elements load dark styles.

How can I do this without first loading the DOM and briefly displaying the light styles? It creates a flash of white before the dark mode is loaded. Because the site is static I can't serve HTML with a dark class already inserted.

Any ideas?
Thank you!

Awesome tool, btw. 🎉

**Comment by richardtallent** [comment]
One possibility would be to start with `<body class="hidden">` (or applied to your `#app` or similar root-level div), then remove that class when the dark or light style has been determined (and the dark class has been set).

Another option is to store the user's preference in `localStorage`, which would allow you to have a tiny bit of JavaScript in your static HTML document that executes before your app module is loaded. That script could add the "dark" class based on the user preference, so it's already set when the app initializes.

**Comment by Nick-Mazuk** [comment]
They key is to both utilize LocalStorage and to take advantage of the event loop and how modern browsers load JavaScript.

To save their theme into local storage, just use this JavaScript:

```js
localStorage.setItem('theme', 'dark')
```

Then, add some JavaScript to the HTML to set the `dark` class accordingly. To prevent the "white flash", it must be in the actual HTML file and not an external JS file. When you do that, here's what the browser will do:

1. Parse the html
2. Run the JS in the script tag
3. Start rendering the page

So if you embed the JS directly into the HTML, you can set the class before the page is rendered.

---

Here's the code snippet I include on my sites. You can just copy/paste it if you want. Note that users can set the theme to three states: `light`, `dark`, and `system`. Here's what it does:

1. If theme is set, add the `dark` class accordingly
2. If the theme is not set, use the system theme

```html
<script>
    const setTheme = () => {
        const savedTheme = localStorage.getItem('theme')
        if (!savedTheme) localStorage.setItem('theme', 'system')
        const { classList } = document.body
        if (savedTheme === 'system' || !savedTheme) {
            const isDark = window.matchMedia('(prefers-color-scheme: dark)').matches
            if (isDark) classList.add('dark')
        } else if (savedTheme === 'dark') {
            classList.add('dark')
        }
    }
    setTheme()
</script>
```

**Reply by johnmathews** [reply]
Thanks for the advice!
I've been using the localstorage approach to set and retrieve settings but hadn't thought of un-hiding the page, and didn't know about the event loop.

Hiding then un-hiding the page doesn't seem to work. I still get a flash, presumably from the entire page being hidden and the default background color being white.

@Nick-Mazuk should the javascript be in the the `<head>` section of the html or <body>? should it be at the end of the body? I've moved the code out of a .js file and into the html, but its still flashing white.

**Reply by johnmathews** [reply]
the js snippet I'm using is:

```js
      <script>
        (function(){
          if (localStorage.getItem('theme') === 'dark') {
                    document.getElementById('baz').classList.add('dark')
                    $("#PygmentCSS").attr('href', dark)
                  }
        })
      </script>
```

Its just before the body tag is closed. Does that sound like it should work?

**Reply by richardtallent** [reply]
Your IIFE isn't called above, so it's not actually executed as written... you need another `()` after the closing `)`. Also, it appears to rely on JQuery being loaded, which is a dependency you should avoid if you're trying to have it run as soon as possible during the page load process.

The best-case solution is for this to run in the `<head>`, but at that point, the only DOM element available is `<html>`. But that's fine, that's still an ancestor of your content, so the dark class can be applied to it.

```HTML
<html>
	<head>
		<script>
			if (localStorage.getItem('theme') === 'dark') {
				document.getElementsByTagName( 'html' )[0].classList.add("dark");
			}
		</script>
	</head>
...
```

An IIFE is not needed since we're not declaring any variables or functions that could clutter your document scope. Anything else that needs to be done based on the dark/light mode setting (such as showing the right menu option for the use to toggle them) should happen later, in your application code.

Since you're using JQuery, am I right to assume you're not using a reactive DOM approach like Vue or React? If that's the case, and your HTML file contains the _literal contents of your page_ rather than just a JavaScript file to bootstrap an SPA, you may need an "onload" event handler as well to unhide your body only when all the page has completely loaded (including the CSS file). That will decrease your FCP metric, especially if you have images or web fonts on the page.

**Comment by johnmathews** [comment]
Thanks for your help and the detailed advice. 

I don't think I can eliminate all the white flashes, at least in Firefox because I'm using a multi page app. 

When Firefox is waiting for a page to load as I navigate from one page to the next, the color of the empty window before the browser gets any data is white. 

Once data is loaded I can control what happens, but there is sometimes (depending on the network I guess) a gap between leaving one page and loading the next page and this is where the window flashes white. In a MPA this wouldn't be a problem. Does that sound correct?

**Comment by lukewarlow** [comment]
One way to eliminate the white flash is to add `<meta name="color-scheme" content="dark">` to the html. This will mean that the browsers default flash will be dark mode not light mode.

**Reply by lnfel** [reply]
Now light mode has the issue of flashing dark lights...
